### PR TITLE
generate systemd: add --start-timeout flag

### PR DESCRIPTION
**Pull request by vrothberg**
Add a new flag to set the start timeout for a generated systemd unit.
To make naming consistent, add a new --stop-timeout flag as well and let
the previous --time map to it.

Fixes: #11618
Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
@containers/podman-maintainers PTAL

**Comment by flouthoc** [comment]
LGTM, just waiting for the flake to pass.

**Comment by umohnani8** [comment]
LGTM

File: docs/source/markdown/podman-generate-systemd.1.md [hunk 1]
```diff
@@ -38,7 +38,11 @@ Note that `--new` only works on containers and pods created directly via Podman
...
 Do not generate the header including meta data such as the Podman version and the timestamp.
 
-#### **--time**, **-t**=*value*
+#### **--start-timeout** =*value*
+
+Override the default start timeout for the container with the given value.
+
+#### **--stop-timeout** =*value*
 
 Override the default stop timeout for the container with the given value.
```
Comment (Luap99) [inline docs/source/markdown/podman-generate-systemd.1.md:47]: Can add that this value is in seconds.

File: pkg/domain/infra/tunnel/generate.go [hunk 2]
```diff
@@ -8,14 +8,11 @@ import (
...
 func (ic *ContainerEngine) GenerateSystemd(ctx context.Context, nameOrID string, opts entities.GenerateSystemdOptions) (*entities.GenerateSystemdReport, error) {
-	options := new(generate.SystemdOptions).WithUseName(opts.Name).WithContainerPrefix(opts.ContainerPrefix).WithNew(opts.New).WithNoHeader(opts.NoHeader).WithTemplateUnitFile(opts.TemplateUnitFile)
+	options := new(generate.SystemdOptions).WithUseName(opts.Name).WithContainerPrefix(opts.ContainerPrefix).WithNew(opts.New).WithNoHeader(opts.NoHeader).WithTemplateUnitFile(opts.TemplateUnitFile).WithStartTimeout(opts.StartTimeout).WithStopTimeout(opts.StopTimeout)
 	options.WithPodPrefix(opts.PodPrefix).WithSeparator(opts.Separator)
 	if opts.RestartPolicy != nil {
 		options.WithRestartPolicy(*opts.RestartPolicy)
 	}
-	if to := opts.StopTimeout; to != nil {
-		options.WithStopTimeout(*opts.StopTimeout)
-	}
```
Comment (Luap99) [inline pkg/domain/infra/tunnel/generate.go:18]: Why did you change this. Now the client will send the default value form containers.conf and it will not use the servers default.

File: pkg/api/server/register_generate.go [hunk 3]
```diff
@@ -37,7 +37,12 @@ func (s *APIServer) registerGenerateHandlers(r *mux.Router) error {
 	//    default: false
 	//    description: Do not generate the header including the Podman version and the timestamp.
 	//  - in: query
-	//    name: time
+	//    name: startTimeout
+	//    type: integer
+	//    default: 0
+	//    description: Start timeout override.
```
Comment (Luap99) [inline pkg/api/server/register_generate.go:43]: Please add that this is in seconds

File: pkg/domain/infra/tunnel/generate.go (hunk 2 above)
Comment (vrothberg) [reply pkg/domain/infra/tunnel/generate.go:18]: Good catch.  I'll fix that.

File: cmd/podman/generate/systemd.go [hunk 4]
```diff
@@ -56,8 +59,17 @@ func init() {
...
-	_ = systemdCmd.RegisterFlagCompletionFunc(timeFlagName, completion.AutocompleteNone)
+	flags.UintVarP(&startTimeout, startTimeoutFlagName, "", 0, "Start timeout override")
+	_ = systemdCmd.RegisterFlagCompletionFunc(startTimeoutFlagName, completion.AutocompleteNone)
+
+	// NOTE: initially, there was only a --time/-t flag which mapped to
+	// stop-timeout. To remain backwards compatible create a hidden flag
+	// that maps to StopTimeout.
+	flags.UintVarP(&stopTimeout, stopTimeoutFlagName, "", containerConfig.Engine.StopTimeout, "Stop timeout override")
+	_ = systemdCmd.RegisterFlagCompletionFunc(stopTimeoutFlagName, completion.AutocompleteNone)
+	flags.UintVarP(&stopTimeout, stopTimeoutCompatFlagName, "t", containerConfig.Engine.StopTimeout, "Backwards alias for --stop-timeout")
```
Comment (Luap99) [inline cmd/podman/generate/systemd.go:70]: Sorry, I forgot to say this yesterday. You should make this an alias instead of a hidden flag, see https://github.com/containers/podman/blob/1bfbb28b0365790552483b961b4bd48a69dd8070/cmd/podman/system/unshare.go#L38-L46

Comment (vrothberg) [reply cmd/podman/generate/systemd.go:70]: Why?  I don't want `--time` to be displayed anymore.

Comment (vrothberg) [reply cmd/podman/generate/systemd.go:70]:
```
$ ./bin/podman generate systemd test --time 1234
Error: unknown flag: --time
```

Comment (vrothberg) [reply cmd/podman/generate/systemd.go:70]: Aaaah, nevermind, I see what's wrong now.

Comment (vrothberg) [reply cmd/podman/generate/systemd.go:70]: How shall we deal with `-t`?  I am unable to catch that in the SetNormalizedFunc.

Comment (Luap99) [reply cmd/podman/generate/systemd.go:70]: Ah I didn't knew that it cannot handle short flag names...
I think we have to stick with your current version, sorry for wasting your time on this.

Comment (vrothberg) [reply cmd/podman/generate/systemd.go:70]: No worries at all.  It was a useful exercise :+1: Thanks for checking!
